crates/pathfinder/src/storage/state.rs

```diff
@@ -799,24 +803,41 @@ impl StarknetEventsTable {
             params.push((":events_match", &key_fts_expression));
         }
 
+        // Paging
+        if filter.page_size > Self::PAGE_SIZE_LIMIT {
+            return Err(EventFilterError::PageSizeTooBig(Self::PAGE_SIZE_LIMIT).into());
+        }
+        if filter.page_size < 1 {
+            anyhow::bail!("Invalid page size");
+        }
+        let offset = filter.page_number * filter.page_size;
+        // We have to be able to decide if there are more events. We request one extra event
+        // above the requested page size, so that we can decide.
+        let limit = filter.page_size + 1;
+        params.push((":limit", &limit));
+        params.push((":offset", &offset));
+
         let query = if where_statement_parts.is_empty() {
             format!(
-                "{} ORDER BY block_number, transaction_hash, idx",
+                "{} ORDER BY block_number, transaction_hash, idx LIMIT :limit OFFSET :offset",
                 base_query
             )
         } else {
             format!(
-                "{} WHERE {} ORDER BY block_number, transaction_hash, idx",
+                "{} WHERE {} ORDER BY block_number, transaction_hash, idx LIMIT :limit OFFSET :offset",
                 base_query,
-                where_statement_parts.join(" AND ")
+                where_statement_parts.join(" AND "),
             )
         };
 
-        let mut statement = connection.prepare(&query)?;
-        let mut rows = statement.query(params.as_slice())?;
+        let mut statement = connection.prepare(&query).context("Preparing SQL query")?;
+        let mut rows = statement
+            .query(params.as_slice())
+            .context("Executing SQL query")?;
 
+        let mut is_last_page = true;
         let mut emitted_events = Vec::new();
-        while let Some(row) = rows.next()? {
+        while let Some(row) = rows.next().context("Fetching next event")? {
```

Comment:
Is it (trivially) possible to precount and preallocate the vector capacity?

Reply:
I haven't found a way to do that without doing an extra "COUNT (*)" query first. Since that query would be quite complicated, too I thought it's probably not worth it.

Also, Vec::push() is supposed to be amortized O(1), which requires handling potential re-allocations in a non-naive way.